### PR TITLE
[Nodegen] Add the possibility to have variable size outputs

### PR DESCRIPTION
**Pull request by tlepley-cadence**
### Summary
Add the possibility to specify a node with variable number of outputs (we already had this feature for inputs but not for outputs)

### Documentation:
This feature is useful in some cases we have encountered.

The most visible change relates the node API change : Node::getOutputName() now returns a std::string instead of a llvm::StringRef.Example This gives the freeom to generate output names on the fly and makes the output API similar to the input API.

Example:
_BB.newBackendSpecificNode("TestNodeVector")
    .addMember(MemberType::VectorNodeValue, "Inputs")
    .addVariableResult("Outputs")
    .setDocstring("Node with variable number of inputs and outputs,")_;

### Test Plan:
This is tested internally since we use this feature. I didn't add a unit test in this PR because this would mean adding a fake node to Glow just for testing purpose. Another solution is to create a backend just for testing this, what seems overkill.

### CODE REVIEW

**Comment by mciprian13** [comment]
@opti-mix @jfix71 Do you know whether can someone look over this PR? I will also need this feature in the near future and it will be good to have the capability to define variable number of inputs/outputs for the Glow nodes. Thanks!

**Comment by tlepley-cadence** [comment]
@mciprian13 I've fixed the model export aspect locally. I'll update and rebase my branch shortly.
This is a feature we have been using multiple times and we find it very useful in our compiler.

**Comment by tlepley-cadence** [comment]
I've just updated the branch with the latest improvements I had done locally. Note that this branch also fixes an issue when exporting ops with variable number of inputs.

**Comment by tlepley-cadence** [comment]
Apparently, the CI is currently broken
```[1/7079] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotobuf.dir/__/src/google/protobuf/any_lite.cc.o
FAILED: third_party/protobuf/cmake/CMakeFiles/libprotobuf.dir/__/src/google/protobuf/any_lite.cc.o 
/usr/local/bin/ccache /tmp/sccache/c++ -DGOOGLE_PROTOBUF_CMAKE_BUILD -DHAVE_PTHREAD -I/tmp/pytorch/build/third_party/protobuf/cmake -I/tmp/pytorch/third_party/protobuf/src -fvisibility-inlines-hidden -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG -fPIC -std=c++11 -MD -MT third_party/protobuf/cmake/CMakeFiles/libprotobuf.dir/__/src/google/protobuf/any_lite.cc.o -MF third_party/protobuf/cmake/CMakeFiles/libprotobuf.dir/__/src/google/protobuf/any_lite.cc.o.d -o third_party/protobuf/cmake/CMakeFiles/libprotobuf.dir/__/src/google/protobuf/any_lite.cc.o -c /tmp/pytorch/third_party/protobuf/src/google/protobuf/any_lite.cc
ccache: error: /var/lib/jenkins/.ccache/ccache.conf: No such file or directory

```

File: tools/ClassGen/NodeBuilder.cpp [hunk 1]
```diff
@@ -213,6 +234,27 @@ void NodeBuilder::emitSettersGetters(std::ostream &os) const {
...
+    for (const auto &op : variableOutputMembers_) {
+      os << "  const std::vector<NodeValue> get" << op.second << "() const {\n"
+         << "    unsigned_t indexShift = " << indexShift << ";\n"
+         << "    std::vector<NodeValue> valueList;\n"
+         << "    for (unsigned i = 0; i < " << op.second << "_size_; i++) {\n"
+         << "      valueList.push_back(getNthResult(indexShift + i));\n"
+         << "    }\n"
+         << "    return valueList;\n"
+         << "  }\n";
+      os << "  const NodeValue get" << op.second << "_Nth(unsigned i) const {\n"
```
[the remark this answers is not in the document]
Comment (opti-mix) [reply tools/ClassGen/NodeBuilder.cpp:249]: It could be useful to add a comment explaining the semantics. If I understand you correctly, this method would return the i-th element of the current variable size output, correct?

File: tools/ClassGen/NodeBuilder.cpp [hunk 2]
```diff
@@ -293,12 +335,23 @@ void NodeBuilder::emitEdges(std::ostream &os) const {
...
-  os << "\nllvm::StringRef " << name_
+  os << "\nstd::string " << name_
      << "Node::getOutputName(unsigned idx) const {\n";
+  // Handle fixed outputs.
   for (size_t i = 0; i < nodeOutputs_.size(); i++) {
     os << "  if (idx == " << i << ") { return \"" << nodeOutputs_[i].second
        << "\"; }\n";
   }
+  // Handle variable outputs.
+  os << "  idx -= " << nodeOutputs_.size() << ";\n";
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.cpp:346]: Is this done to use nodeOutput-relative indexing?

File: tools/ClassGen/NodeBuilder.cpp [hunk 3]
```diff
@@ -352,23 +405,53 @@ void NodeBuilder::emitPrettyPrinter(std::ostream &os) const {
...
     os << "  db.addParam(\"" << op.second << "\", *(get" << op.second
        << "().getType()));\n";
   }
+  // --> variable outputs.
+  if (variableOutputMembers_.size()) {
+    os << "  unsigned_t variableOutIdx = " << nodeOutputs_.size() << ";\n";
+    for (const auto &op : variableOutputMembers_) {
+      os << "  for (unsigned i = 0; i < " << op.second << "_size_; i++) {\n"
+         << "   db.addParam(\"" << op.second << "\" + std::to_string(i),"
+         << " *(getNthResult(variableOutIdx + i).getType()));\n"
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.cpp:419]: Could you add comments? Are you trying to get Type of the i-th element of op.second variable size output? If so, maybe it is safer to call "get + op.second + NthResult(i)" instead of hard-coding the current mapping via a general getNthResult?

File: tools/ClassGen/NodeBuilder.cpp [hunk 4]
```diff
@@ -352,23 +405,53 @@ void NodeBuilder::emitPrettyPrinter(std::ostream &os) const {
...
 }
 
 void NodeBuilder::emitCloner(std::ostream &os) const {
   os << "\nNode* " << name_ << "Node::clone() const {\n";
 
+  // Variable outputs
+  if (variableOutputMembers_.size()) {
+    os << "  unsigned_t variableOutIdx = " << nodeOutputs_.size() << ";\n";
+    for (const auto &op : variableOutputMembers_) {
+      os << "  std::vector<TypeRef> ResultTypes_of_" << op.second << ";\n"
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.cpp:435]: We usually don't use underscores.

Comment (opti-mix) [reply tools/ClassGen/NodeBuilder.cpp:435]: Nit: Since Node has "types_" vector, couldn't you just use an ArrayRef here which would just be a slice of the "types_" array corresponding to the variable output? This way you don't need to copy anything around.

Or do we need to pass it as a  vector later?

File: tools/ClassGen/NodeBuilder.h [hunk 5]
```diff
@@ -46,6 +46,7 @@ class NodeBuilder {
   std::vector<std::pair<std::string, std::string>> nodeOutputs_;
   /// A list of node members. Format: (type, name).
   std::vector<std::pair<MemberTypeInfo, std::string>> members_;
+  std::vector<std::pair<MemberTypeInfo, std::string>> variableOutputMembers_;
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.h:49]: Please add  a Doxygen comment for this member.

**Review by opti-mix — COMMENTED** [review]
Overall, the PR looks reasonable. 

But if possible, could you add in the PR comment an example of how you define a node with e.g. 2 or 3 variable size outputs and what code is generated from it? Without it, it is a bit hard to check some parts of this PR for correctness. Thanks!

**Comment by mciprian13** [comment]
@tlepley-cadence We have started testing this PR. We will let you know if we find any problems. Meanwhile could you please address the review items raised by opti-mix so we can land this? Thanks!

**Comment by tlepley** [comment]
@mciprian13 I was busy with other things this week, but I'll try to find some time early next week to address comment. This PR has been around for some time and I think it would be good to land it ASAP.

**Comment by mciprian13** [comment]
@tlepley-cadence While playing with this PR we discovered that this PR only handles the variable output size for the nodes only. The question is how to have also instructions with variable number of input/output operands while also doing automating IRGen with the ```.autoIRGen()``` flag?

**Comment by tlepley-cadence** [comment]
@mciprian13 like specified in the PR title, this PR indeed only tacked variable outputs for the _nodegen_ part, not the _instrgen_ part.

In our case, most of the usage a variable outputs is needed on the node level. We only have one case where we need variable outputs at the instruction level and for now, our approach has been to define a 'custom' instruction (not using _instrgen_).

In order to convert the op with variable outputs (generated by nodegen) to a 'custom' instruction with variable outputs, the solution is today to do the translation manually in iIRGen.cpp. There are usually some manual translations happening there even for some nodegen created nodes.

So in conclusion, this PR saves you the burden of handling variable outputs on the node part, but not for the instruction part. Some effort would be needed there but we didn't find the need critical enough for now.

File: tools/ClassGen/NodeBuilder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp:249]: Yes this is correct. We need such methods because there can be multiple variable outputs. the introduction of '_' is to better separate the output name (given by the user) from the functionality.

File: tools/ClassGen/NodeBuilder.cpp [hunk 6]
```diff
@@ -182,6 +198,10 @@ void NodeBuilder::emitClassMembers(std::ostream &os) const {
   for (const auto &op : members_) {
     os << "  " << getStorageTypename(&op.first) << " " << op.second << "_;\n";
   }
+
+  for (const auto &op : variableOutputMembers_) {
+    os << "  unsigned " << op.second << "_size_;\n";
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp:203]: I have added '_' to well separate the member name given by the user and then to limit the risk of collision with member names. I can remove these '_' to follow the usual Glow convention.

File: tools/ClassGen/NodeBuilder.cpp (hunk 2 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp:346]: Yes, It indeed makes _idx_ relative to the current variable output. From this, it's simpler to detect of the index corresponds to a value of this particular output and to generate a name for the value (that contains the relative index).

File: tools/ClassGen/NodeBuilder.cpp (hunk 3 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp:419]: yes, the type of i-th element of op.second variable size output.
On don't think it's sager since `getNthResult` will also manage the variability properly, but it can be more consistent with the rest of the code (that gets the size with the output name).
It would be : `"get" << op.second << "Nth(i)"`

File: tools/ClassGen/NodeBuilder.cpp (hunk 4 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp:435]: I think `ArrayRef `is fine. `Node` class `type_` member is protected we we may indeed use `ArrayRef` to avoid creating vectors on the fly.

File: tools/ClassGen/NodeBuilder.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Removed `_` to stick the the Glow naming convention

File: tools/ClassGen/NodeBuilder.cpp [hunk 7]
```diff
@@ -213,6 +234,27 @@ void NodeBuilder::emitSettersGetters(std::ostream &os) const {
...
+         << "    std::vector<NodeValue> valueList;\n"
+         << "    for (unsigned i = 0; i < " << op.second << "_size_; i++) {\n"
+         << "      valueList.push_back(getNthResult(indexShift + i));\n"
+         << "    }\n"
+         << "    return valueList;\n"
+         << "  }\n";
+      os << "  const NodeValue get" << op.second << "_Nth(unsigned i) const {\n"
+         << "    return getNthResult(" << indexShift << " + i);\n"
+         << "  }\n";
+      os << "  unsigned get" << op.second << "_Size() const {\n"
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Removed `_` to stick the the Glow naming convention

File: tools/ClassGen/NodeBuilder.cpp (hunk 3 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Comment added and use the specific value getter rather than the generic one.

File: tools/ClassGen/NodeBuilder.h (hunk 5 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.h]: Done.

**Comment by tlepley-cadence** [comment]
@opti-mix I've addressed your feedbacks. Please take a look.

File: tools/ClassGen/NodeBuilder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Removed `_` to stick the the Glow naming convention

File: tools/ClassGen/NodeBuilder.cpp (hunk 4 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Removed `_` to stick the the Glow naming convention.
Also used `ArrayRef` slices to avoid building vectors on the fly.

**Comment by tlepley-cadence** [comment]
The CI looks broken:
```
../tests/benchmark/SLSBench.cpp:25:10: fatal error: 'glow/glow/tests/benchmark/CommonSLSEB.h' file not found
#include "glow/glow/tests/benchmark/CommonSLSEB.h"
```

File: tools/ClassGen/NodeBuilder.cpp [hunk 8]
```diff
@@ -352,23 +410,51 @@ void NodeBuilder::emitPrettyPrinter(std::ostream &os) const {
...
 
 void NodeBuilder::emitCloner(std::ostream &os) const {
   os << "\nNode* " << name_ << "Node::clone() const {\n";
 
+  // Variable outputs
+  if (variableOutputMembers_.size()) {
+    os << "  unsigned_t variableOutIdx = " << nodeOutputs_.size() << ";\n"
+       << "  llvm::ArrayRef<TypeRef> typeArray(types_);\n";
+    for (const auto &op : variableOutputMembers_) {
+      os << "  llvm::ArrayRef<TypeRef> ResultTypesOf" << op.second
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.cpp:441]: ResultTypesOf should be camelCase, i.e. resultTypesOf.

File: tools/ClassGen/NodeBuilder.cpp [hunk 9]
```diff
@@ -624,6 +723,22 @@ void NodeBuilder::emitImportMethods(std::ostream &os) const {
          << ", dict));\n\n";
     }
   }
+  for (const auto &op : variableOutputMembers_) {
+    auto ty = getCtorArgTypename(MemberType::Unsigned);
+    os << "  " << ty << " " << op.second << "_Size;\n"
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.cpp:728]: We usually don't use underscores in the middle of a name.

File: tools/ClassGen/NodeBuilder.cpp [hunk 10]
```diff
@@ -624,6 +723,22 @@ void NodeBuilder::emitImportMethods(std::ostream &os) const {
...
+       << "   std::vector<TypeRef> " << op.second << "_OutTyList;\n"
```
Comment (opti-mix) [inline tools/ClassGen/NodeBuilder.cpp:729]: Same here.

**Comment by opti-mix** [comment]
@tlepley-cadence Thanks for addressing the review comments. I had a look and provided a couple more cosmetic comments.

I'd also like to ask again for the following:
if possible, could you add in a reply to this message an example of how you define a node with e.g. 2 or 3 variable size outputs and what code is generated from it? Without it, it is a bit hard to check some parts of this PR for correctness. Thanks!

I haven't checked the ONNX part as I don't know it well enough. @jfix71 could you have a look?

File: tools/ClassGen/NodeBuilder.cpp [hunk 11]
```diff
@@ -675,25 +793,59 @@ void NodeBuilder::emitExportMethods(std::ostream &os) const {
...
+    std::string sParamSize = "N__->get" + op.second + "().size()";
+    os << "    addValueAttribute(opProto, \"" << op.second
+       << "_Index\", index);\n";
+    os << "    addValueAttribute(opProto, \"" << op.second << "_Size\", "
+       << sParamSize << ");\n";
+    os << "    index += " + sParamSize + ";\n";
+  }
+  if (hasVariadicInput) {
+    os << "  }\n";
+  }
```
Comment (jfix71) [inline tools/ClassGen/NodeBuilder.cpp:835]: Perhaps I'm missing something -- why do we need to export this extra variadic inputs info here w/ Index and Size? We already handle variadic inputs when we write out node members (e.g. for nodes like concat) here, right?

https://github.com/pytorch/glow/blob/61da45dfd1a7a563b68de5584c4a2237b1ce9d47/tools/ClassGen/NodeBuilder.cpp#L699-L712

**Review by jfix71 — COMMENTED** [review]
Added a question and a few nits. 

Another question -- is there a concern here for ONNX backwards compatibility for variadic inputs? E.g. if we try to load a previously exported ONNX model that has a concat, it would still work, correct?

File: tools/ClassGen/NodeBuilder.cpp (hunk 9 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Fixed.

File: tools/ClassGen/NodeBuilder.cpp (hunk 10 above)
Comment (tlepley-cadence) [reply tools/ClassGen/NodeBuilder.cpp]: Fixed.

**Comment by opti-mix** [comment]
@tlepley-cadence 

Another try ;-)

if possible, could you add in a reply to this message an example of how you define a node with e.g. 2 or 3 variable size outputs and what code is generated from it? Without it, it is a bit hard to check some parts of this PR for correctness. Thanks!

**Comment by khabinov** [comment]
@opti-mix yup, I will check it and report back today or tomorrow if that's ok.

**Comment by mciprian13** [comment]
@tlepley-cadence You mentioned in the PR description that this feature already existed for the inputs but I don't see a similar method `addVariableInput` in NodeGen. Can you tell me how is that done? Thank you!
